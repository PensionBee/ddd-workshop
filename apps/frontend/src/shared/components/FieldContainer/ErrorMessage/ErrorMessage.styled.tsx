import { Box } from "@shared/components/Box";
import { theme } from "@shared/styling/theme";
import styled, { css } from "styled-components";

const errorMessageArrowStyles = css`
  content: "";
  position: absolute;
  top: -10px;
  left: 16px;
  border: 5px solid transparent;
  border-bottom-color: ${() => theme.colors.red100};
`;

export const StyledErrorMessage = styled(Box)`
  &:before {
    ${errorMessageArrowStyles}
  }
`;
