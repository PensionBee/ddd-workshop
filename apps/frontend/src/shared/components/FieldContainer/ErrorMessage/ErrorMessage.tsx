import { DisclaimerR } from "@shared/components/Typography";
import { StyledErrorMessage } from "./ErrorMessage.styled";

export type ErrorMessageProps = {
  /** Error message to be displayed */
  children?: string;
};

export const ErrorMessage: React.FC<ErrorMessageProps> = ({ children }) => {
  if (!children) return null;
  return (
    <StyledErrorMessage
      px={16}
      py={4}
      backgroundColor="red100"
      borderRadius={4}
      mt={5}
    >
      <DisclaimerR color="white">{children}</DisclaimerR>
    </StyledErrorMessage>
  );
};
