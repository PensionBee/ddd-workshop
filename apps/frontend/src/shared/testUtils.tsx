import { render } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import { ThemeProvider } from "styled-components";
import type { Breakpoints } from "./styling/breakpoints";
import { BREAKPOINTS } from "./styling/breakpoints";
import { theme } from "./styling/theme";
import { vi } from "vitest"

type ReactTestingLibraryProviderProps = {
  children: React.ReactNode;
};

// eslint-disable-next-line react-refresh/only-export-components
const ReactTestingLibraryProviders: React.FC<
  ReactTestingLibraryProviderProps
> = ({ children }) => {
  return <ThemeProvider theme={theme}>{children}</ThemeProvider>;
};

// eslint-disable-next-line @typescript-eslint/no-explicit-any
const customRender = (ui: React.ReactElement, options?: any) => {
  return render(ui, { wrapper: ReactTestingLibraryProviders, ...options });
};

// re-export everything
// eslint-disable-next-line react-refresh/only-export-components
export * from "@testing-library/react";

// override render method
export { customRender as render };

export const sleep = (delay: number) => {
  return new Promise((resolve) => setTimeout(resolve, delay));
};

export const mockMediaMatches = (breakpoint: Breakpoints) => {
  Object.defineProperty(window, "matchMedia", {
    writable: true,
    value: (query: string) => {
      const breakpointValue = BREAKPOINTS[breakpoint];
      const queryValue = parseInt(query.replace(/^\D+/g, ""));
      const breakpointValues = [...Object.values(BREAKPOINTS)];
      if (!breakpointValues.includes(queryValue)) {
        throw new Error(
          `Invalid breakpoint value: ${queryValue}, this likely means there has been a change to the way that breakpoints are setup.`
        );
      }
      return {
        matches: queryValue <= breakpointValue,
        addEventListener: vi.fn(),
        removeEventListener: vi.fn(),
      };
    },
  });
};

export const routerRender = (ui: React.ReactNode, pathname = "/") => {
  return customRender(
    <MemoryRouter initialEntries={[{ pathname }]}>{ui}</MemoryRouter>
  );
};
