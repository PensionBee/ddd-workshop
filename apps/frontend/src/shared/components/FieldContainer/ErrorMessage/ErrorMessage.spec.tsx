import { render, vi } from "@shared/testUtils";
import { ErrorMessage } from "./ErrorMessage";

const ERROR_MESSAGE = "Error message";

describe("FieldWrapper/ErrorMessage", () => {
  it("should render correctly", () => {
    const { getByText, asFragment } = render(
      <ErrorMessage>{ERROR_MESSAGE}</ErrorMessage>
    );

    expect(getByText(ERROR_MESSAGE)).toBeVisible();
    expect(asFragment()).toMatchSnapshot();
  });
  it("should not render if no children", () => {
    const { container } = render(<ErrorMessage />);

    expect(container.firstChild).toBeNull();
  });
});
