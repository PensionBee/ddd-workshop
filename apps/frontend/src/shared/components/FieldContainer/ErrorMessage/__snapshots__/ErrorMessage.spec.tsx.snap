// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`FieldWrapper/ErrorMessage > should render correctly 1`] = `
<DocumentFragment>
  <div
    class="sc-gsnScm cUcvaQ sc-dkzCjb kZxHOV"
  >
    <span
      class="sc-bczSft imQpyU"
      data-rc="DisclaimerR"
    >
      Error message
    </span>
  </div>
</DocumentFragment>
`;

exports[`FieldWrapper/ErrorMessage should render correctly 1`] = `
<DocumentFragment>
  <div
    class="sc-gswNZR bwgTNC sc-dkrFOg kvDetP"
  >
    <span
      class="sc-bcXHqe hVJkQo"
      data-rc="DisclaimerR"
    >
      Error message
    </span>
  </div>
</DocumentFragment>
`;
